Add options to filter lists_mergeby (#4058)
ansible-collections/community.general: docs/docsite/rst/filter_guide.rst

Merging lists of dictionaries
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If you have two or more lists of dictionaries and want to combine them into a list of merged dictionaries, where the dictionaries are merged by an attribute, you can use the ``lists_mergeby`` filter.

.. note:: The output of the examples in this section use the YAML callback plugin. Quoting: "Ansible output that can be quite a bit easier to read than the default JSON formatting." See :ref:`the documentation for the community.general.yaml callback plugin <ansible_collections.community.general.yaml_callback>`.

In the example below the lists are merged by the attribute ``name``:

.. literalinclude:: examples/lists_mergeby/example-001.yml
   :language: yaml+jinja

This produces:

.. literalinclude:: examples/lists_mergeby/example-001.out
   :language: yaml

.. versionadded:: 2.0.0

It is possible to use a list of lists as an input of the filter:

.. literalinclude:: examples/lists_mergeby/example-002.yml
   :language: yaml+jinja

This produces the same result as in the previous example:

.. literalinclude:: examples/lists_mergeby/example-002.out
   :language: yaml

> The filter also accepts two optional parameters: ``recursive`` and ``list_merge``. These parameters are only supported when used with ansible-base 2.10 or ansible-core, but not with Ansible 2.9.

Reviewer: ```suggestion
The filter also accepts two optional parameters: ``recursive`` and ``list_merge``. These parameters are only supported when used with ansible-base 2.10 or ansible-core, but not with Ansible 2.9. This is available since community.general 4.4.0.
```

**recursive**
    Is a boolean, default to ``False``. Should the ``combine`` recursively merge nested hashes. Note: It does not depend on the value of the ``hash_behaviour`` setting in ``ansible.cfg``.

**list_merge**
    Is a string, its possible values are ``replace`` (default), ``keep``, ``append``, ``prepend``, ``append_rp`` or ``prepend_rp``. It modifies the behaviour of community.general.lists_mergeby when the hashes to merge contain arrays/lists.

> The examples below set ``recursive=True`` and display the differences among all six options of ``list_merge``. Functionality of the parameters is exactly the same as in the filter ``combine``. See `Combining hashes/dictionaries <https://docs.ansible.com/ansible/devel/user_guide/playbooks_filters.html#combining-hashes-dictionaries>`_ to learn details about these options.

Reviewer: ```suggestion
The examples below set ``recursive=true`` and display the differences among all six options of ``list_merge``. Functionality of the parameters is exactly the same as in the filter ``combine``. See `Combining hashes/dictionaries <https://docs.ansible.com/ansible/devel/user_guide/playbooks_filters.html#combining-hashes-dictionaries>`_ to learn details about these options.
```

> Example **list_merge=replace** (default)

Reviewer: ```suggestion
Example **list_merge=replace** (default):
```
Please also add this to the others.

Merging lists of dictionaries
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If you have two or more lists of dictionaries and want to combine them into a list of merged dictionaries, where the dictionaries are merged by an attribute, you can use the ``lists_mergeby`` filter.

> .. note:: The output of the examples in this section use the yaml callback plugin. Quoting: "Ansible output that can be quite a bit easier to read than the default JSON formatting." See ``shell> ansible-doc -t callback yaml``

Reviewer: ```suggestion
.. note:: The output of the examples in this section use the YAML callback plugin. Quoting: "Ansible output that can be quite a bit easier to read than the default JSON formatting." See :ref:`the documentation for the community.general.yaml callback plugin <ansible_collections.community.general.yaml_callback>`.
```

> In the example below the lists are merged by the attribute *name*

Reviewer: ```suggestion
In the example below the lists are merged by the attribute *name*:
```

.. literalinclude:: ../examples/lists_mergeby/example-001.yml

>    :language: yaml

Reviewer: The file is not pure YAML. Generally we always use `yaml+jinja` for something that can potentially contain Jinja2.
```suggestion
   :language: yaml+jinja
```

This produces:

.. literalinclude:: ../examples/lists_mergeby/example-001.out
   :language: yaml

.. versionadded:: 2.0.0

> It is possible to use a list of lists as an input of the filer

Reviewer: ```suggestion
It is possible to use a list of lists as an input of the filer:
```

.. literalinclude:: ../examples/lists_mergeby/example-002.yml

>    :language: yaml

Reviewer: ```suggestion
   :language: yaml+jinja
```

This produces the same result as in the previous example:

.. literalinclude:: ../examples/lists_mergeby/example-002.out
   :language: yaml

> .. versionadded:: 4.4.0

Reviewer: This only makes sense if you document the new options. The above does not do that, so it should not have 4.4.0 as `versionadded`.

Author: I'll remove it. Thank you.

      list1:
        - name: foo
          extra: true
        - name: bar
          extra: false
        - name: meh
          extra: true
      list2:
        - name: foo
          path: /foo
        - name: baz
          path: /baz
  - debug:
      var: list3

This produces the same result as in the previous example:

.. code-block:: yaml

  list3:
  - extra: false
    name: bar
  - name: baz
    path: /baz
  - extra: true
    name: foo
    path: /foo
  - extra: true
    name: meh

The filter also accepts two optional parameters: ``recursive`` and ``list_merge``. These parameters are only supported when used with ansible-base 2.10 or ansible-core, but not with Ansible 2.9. This is available since community.general 4.4.0.

**recursive**
    Is a boolean, default to ``False``. Should the ``combine`` recursively merge nested hashes. Note: It does not depend on the value of the ``hash_behaviour`` setting in ``ansible.cfg``.

**list_merge**
    Is a string, its possible values are ``replace`` (default), ``keep``, ``append``, ``prepend``, ``append_rp`` or ``prepend_rp``. It modifies the behaviour of community.general.lists_mergeby when the hashes to merge contain arrays/lists.

> The examples below set ``recursive=true`` and display the differences among all six options of ``list_merge``. Functionality of the parameters is exactly the same as in the filter ``combine``. See `Combining hashes/dictionaries <https://docs.ansible.com/ansible/devel/user_guide/playbooks_filters.html#combining-hashes-dictionaries>`_ to learn details about these options.

Reviewer: ```suggestion
The examples below set ``recursive=true`` and display the differences among all six options of ``list_merge``. Functionality of the parameters is exactly the same as in the filter ``combine``. See :ref:`Combining hashes/dictionaries <combine_filter>` to learn details about these options.
```
